MINIFICPP-1088 - clean up minifiexe and MINIFI_HOME logic

bakaid opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

bakaid commented on main/MiNiFiMain.cpp:
```diff
@@ -142,94 +159,39 @@ int main(int argc, char **argv) {
-			return -1;
-		}
+	// chdir to MINIFI_HOME
+	if (!utils::Environment::setCurrentWorkingDirectory(minifiHome.c_str())) {
```
This will solve the problem of relative paths being relative to the cwd, instead of MINIFI_HOME. A lot of code has been written under the assumption that this is true, but with the Windows service case it is explicitly not (the default working directory is C:\Windows\System32

arpadboda replied:
Will run some tests, but looks good at first

bakaid commented on main/MiNiFiMain.cpp:
```diff
@@ -142,94 +159,39 @@ int main(int argc, char **argv) {
 		return -1;
 	}
 #endif
-	// assumes POSIX compliant environment
```
The original order of sources for MINIFI_HOME here was
for *nix: MINIFI_HOME env -> realpath(argv[0]) (path of the current executable) -> cwd
for Windows: MINIFI_HOME env -> GetModuleHandle (path of the current executable) -> cwd
This has been cleaned up and unified in `determineMinifiHome` as:
 MINIFI_HOME env -> path of the current executable -> cwd
The method used for determining the path of the current executable used now is superior to `realpath(argv[0])`

bakaid commented:
@arpadboda This code cleans up the minifi main code, MINIFI_HOME determination and Properties a bit. It also adds more verbose and specific logging, and a log backend to log to the Windows Event Log which gets used in the Windows Service until the logger is configured from minifi-log.properties, so we have information about the MINIFI_HOME determination logic even when we are running as a service, and without the need to attach a debugger.

@amarmer as this is a Windows-heavy PR, it would be nice if you could look at that parts.

am-c-p-p commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+    // GetEnvironmentVariableA does not set last error to 0 on success, so an error from a pervious API call would influence the GetLastError() later,
+    // so we set the last error to 0 before calling
+    SetLastError(ERROR_SUCCESS);
+    uint32_t ret = GetEnvironmentVariableA(name, buffer.data(), buffer.size());
```
nit - can be used GetEnvironmentVariable.

bakaid replied:
I have been bitten before by a rouge header appearing that included `Windows.h` with `UNICODE` defined, making the macros use the W versions of functions, causing segfaults when called with `char*` strings, so I prefer to explicitly define either the ANSI or the Unicode version of the winapi functions. I think it improves readability as well.

am-c-p-p commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+    if (!overwrite && Environment::getEnvironmentVariable(name).first) {
+      success = true;
+    } else {
+      success = SetEnvironmentVariableA(name, value);
```
nit - can used SetEnvironmentVariable.

am-c-p-p commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+
+  Environment::accessEnvironment([&success, name](){
+#ifdef WIN32
+    success = SetEnvironmentVariableA(name, nullptr);
```
nit - can used SetEnvironmentVariable.

am-c-p-p commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+    // the return value specifies the required size of the buffer,
+    // in characters, including the null-terminating character."
+    while (true) {
+      len = GetCurrentDirectoryA(buffer.size(), buffer.data());
```
nit - can be used GetCurrentDirectory.

am-c-p-p commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+
+  Environment::accessEnvironment([&success, directory](){
+#ifdef WIN32
+    success = SetCurrentDirectoryA(directory);
```
nit - can be used SetCurrentDirectory.

am-c-p-p commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+
+  Environment::accessEnvironment([&cwd](){
+#ifdef WIN32
+    uint32_t len = 0U;
```
This works, but simpler would be:
char dir[MAX_PATH+1] ;
GetCurrentDirectory( sizeof(dir), dir ) ;

am-c-p-p commented on libminifi/src/utils/file/PathUtils.cpp:
```diff
@@ -48,6 +56,35 @@ bool PathUtils::getFileNameAndPath(const std::string &path, std::string &filePat
 
+std::string PathUtils::getFullPath(const std::string& path) {
+#ifdef WIN32
+  std::vector<char> buffer(MAX_PATH);
```
MAX_PATH just 260, could be used `char buffer[MAX_PATH+1];`

am-c-p-p commented on libminifi/src/utils/file/PathUtils.cpp:
```diff
@@ -48,6 +56,35 @@ bool PathUtils::getFileNameAndPath(const std::string &path, std::string &filePat
+  std::vector<char> buffer(MAX_PATH);
+  uint32_t len = 0U;
+  while (true) {
+      len = GetFullPathNameA(path.c_str(), buffer.size(), buffer.data(), nullptr /*lpFilePart*/);
```
nit - - can be used GetFullPathName.

am-c-p-p commented:
I'll continue reviewing tomorrow.

arpadboda commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+    char* path = nullptr;
+    while (true) {
+      path = getcwd(buffer.data(), buffer.size());
+      if (path == nullptr) {
```
This seems to be a bit overcomplicated

```    
    while (true) {
      path = getcwd(buffer.data(), buffer.size());
      if (path != nullptr) {
        cwd = path;
        break;
      } else if (errno == ERANGE) {
        buffer.resize(buffer.size() * 2);
      } else {
        break;  // error
      }
   }
#endif
```

bakaid replied:
The original code reflected my mental reasoning, but this is more more concise, changed, thanks.

bakaid commented on libminifi/src/core/logging/Logger.cpp:
```diff
@@ -0,0 +1,149 @@
+      logger_level = spdlog::level::level_enum::err;
+      break;
+    case info:
+      break;
```
Nowhere, I copied it verbatim: https://github.com/apache/nifi-minifi-cpp/blob/master/libminifi/include/core/logging/Logger.h#L203
logger_level is set to info before the switch-case. I don't like it too much either, but it's the original code.

arpadboda replied:
Oh, god, someone must have skipped the lesson where default case was explained...

bakaid commented on libminifi/src/core/logging/Logger.cpp:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (!delegate_->should_log(logger_level)) {
```
Agreed, although this is not new code.

arpadboda commented on libminifi/test/unit/FileUtilsTests.cpp:
```diff
@@ -146,3 +148,33 @@ TEST_CASE("TestFileUtils::create_dir", "[TestCreateDir]") {
+  REQUIRE(tempDir1 == utils::file::PathUtils::getFullPath(tempDir1));
+  REQUIRE(tempDir1 == utils::file::PathUtils::getFullPath("test1"));
+  REQUIRE(tempDir1 == utils::file::PathUtils::getFullPath("./test1"));
+  REQUIRE(tempDir1 == utils::file::PathUtils::getFullPath("././test1"));
```
Is this expected to work on Win?

bakaid replied:
Yep, and it does.

arpadboda commented on main/MiNiFiMain.cpp:
```diff
@@ -142,94 +159,39 @@ int main(int argc, char **argv) {
 #else
 	if (signal(SIGINT, sigHandler) == SIG_ERR || signal(SIGTERM, sigHandler) == SIG_ERR || signal(SIGPIPE, SIG_IGN) == SIG_ERR) {
-		std::cerr << "Cannot install signal handler" << std::endl;
+    logger->log_error("Cannot install signal handler");
```
This looks very strangely indented.

bakaid replied:
That's what you get for editing in Notepad++ in Windows. Thanks.

bakaid commented on libminifi/test/unit/EnvironmentUtilsTests.cpp:
```diff
@@ -0,0 +1,154 @@
+    std::string value;
+    std::tie(isset, value) = utils::Environment::getEnvironmentVariable(env_name.c_str());
+    if (isset) {
+      std::cerr << env_name << " is set to " << value << std::endl;
```
I ain't no oracle, so I can't tell the results of random operations on multiple threads. The only real thing I'm concerned about is this not segfaulting (or triggering ASAN/valgrind). The outcome is not deterministic.

arpadboda replied:
Fair enough, thanks!

bakaid commented on libminifi/src/core/logging/Logger.cpp:
```diff
@@ -0,0 +1,149 @@
+
+LogBuilder::~LogBuilder() {
+  if (!ignore)
+    log_string(level);
```
Is it? As far as I see LogBuilder has no parent or child classes, or virtual functions for that matter.

arpadboda replied:
My bad, it's virtual is BaseLogger, this code is correct.

bakaid commented on libminifi/src/utils/Environment.cpp:
```diff
@@ -0,0 +1,192 @@
+    // in characters, including the null-terminating character."
+    while (true) {
+      len = GetCurrentDirectoryA(buffer.size(), buffer.data());
+      if (len > buffer.size()) {
```
Yep, and you actually caught a potential bug. On `len == buffer.size()` we should retry as well, as on insufficient space it returns the required length of the buffer INCLUDING the terminating \0, on success however it returns the length of the string WITHOUT the terminating \0;

bakaid commented:
@arpadboda Thanks for the review, addressed your review comments, also updated the MessageCompiler finder cmake magic, let's wait for the AppVeyor to see if it works now.

bakaid commented on libminifi/src/core/logging/WindowsEventLogSink.cpp:
```diff
@@ -0,0 +1,95 @@
+namespace internal {
+
+WORD windowseventlog_sink::type_from_level(const spdlog::details::log_msg& msg) const {
+  switch (static_cast<int>(msg.level)) {
```
As you said, we don't use spdlog::level::off, it is not well documented either what it should be mapped to, so I've decided to log all unexpected/potential future unknown enums at error log level.

bakaid replied:
Completely agree, I prefer that solution as well, but we should clean up our warnings a bit first, because currently we are not warning free even with our current set of warnings, and any new warning would be most likely lost in the current ones. I wouldn't put this on -Werror (or only just in libminifi), because that could break third party extensions.

bakaid commented on libminifi/test/TestBase.cpp:
```diff
@@ -18,6 +18,21 @@
+    adjusted_name = name.substr(clazz.length(), name.length() - clazz.length());
+  if (config && config->shortenClassNames()) {
+    utils::ClassUtils::shortenClassName(adjusted_name, adjusted_name);
+  }
```
This has just been moved from the h to the cpp file to avoid some include issues. You are welcome to create a refactor task for this, if it is used in more places I would agree that a good solution.

szaszm replied:
I didn't realize that it's only been moved. Nevermind then.

bakaid commented on libminifi/include/core/logging/WindowsEventLogSink.h:
```diff
@@ -0,0 +1,70 @@
+}
+}
+}
+}
```
Good catch, will fix.

bakaid commented on libminifi/include/core/logging/WindowsEventLogSink.h:
```diff
@@ -0,0 +1,70 @@
+  public:
+   windowseventlog_sink(const std::string& source_name = "ApacheNiFiMiNiFi");
+
+   virtual ~windowseventlog_sink();
```
I shouldn't have used override in libminifi at all. I copied the class from spdlog's syslog sink, which used it this way. Will fix.

am-c-p-p commented:
Looks good. I'll test it.

bakaid commented:
Addressed the new set of review comments, Travis builds seems fixed. Will wait for others to test this.

am-c-p-p commented:
I built installer, installed, server was running, log created under {installedDir}/Log. 
LGTM.
